src/AWSS3.jl

```diff
@@ -471,52 +474,63 @@ end
 
 
 """
-    s3_list_objects([::AWSConfig], bucket, [path_prefix])
+    s3_list_objects([::AWSConfig], bucket, [path_prefix]; delimiter="/", max_items=1000)
 
 [List Objects](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGET.html)
 in `bucket` with optional `path_prefix`.
 
-Returns `Vector{Dict}` with keys `Key`, `LastModified`, `ETag`, `Size`,
+Returns an iterator of `Dict`s with keys `Key`, `LastModified`, `ETag`, `Size`,
 `Owner`, `StorageClass`.
 """
-function s3_list_objects(aws::AWSConfig, bucket, path_prefix="")
+function s3_list_objects(aws::AWSConfig, bucket, path_prefix=""; delimiter="/", max_items=nothing)
+    return Channel() do chnl
+        more = true
+        num_objects = 0
+        marker = ""
 
-    more = true
-    objects = []
-    marker = ""
-
-    while more
-
-        q = SSDict()
-        if path_prefix != ""
-            q["delimiter"] = "/"
-            q["prefix"] = path_prefix
-        end
-        if marker != ""
-            q["marker"] = marker
-        end
-
-        @repeat 4 try
-
-            r = s3(aws, "GET", bucket; query = q)
-
-            more = r["IsTruncated"] == "true"
-            # FIXME return an iterator to allow streaming of truncated results!
+        while more
+            q = Dict{String, String}()
+            if path_prefix != ""
+                q["prefix"] = path_prefix
+            end
+            if delimiter != ""
+                q["delimiter"] = delimiter
+            end
+            if marker != ""
+                q["marker"] = marker
+            end
```

Comment:
Personal preference but this may be easier to read:
```julia
!isempty(path_prefix) && (q["prefix"] = path_prefix)
!isempty(delimiter) && (q["delimiter"] = delimiter)
!isempty(marker) && (q["marker"] = marker)
```

Reply:
I'd tried to stick with the original convention in this function. I tend to agree with your preferred syntax, but I think we should change the convention everywhere in 1 PR.